### PR TITLE
support ulong values for Json

### PR DESCRIPTION
**Pull request by IgorStepanov**
Hello, Vibe.D!

I've written about my problem on forum: http://forum.rejectedsoftware.com/groups/rejectedsoftware.vibed/thread/25482/

I use vibe.d on small commercial product, and two days ago I got this trouble.
One REST server (api.shodan.io) sent me a JSON like `{..., serial: 17559991181826658461, ...}`
I read the official JSON standard: http://json.org/
Unfortunately, it doesn't define bounds for a number value, thus `99999999999999999999999999` is a correct JSON value.
However the most real systems uses 64-bit (as maximum) signed and unsigned integers for describe integral values.
Thus, supporting of ulong is very needed for users.

I've added a new `Type` value: `uint_`, which represents `ulong` values.
I prefer `int_` type instead of `uint_` when it possible.
If the integral value x <= long.max, I write is to Json as `int_` even if it hasn't the sign.

If this way is unacceptable, please suggest another way to parse ulong values.
We may convert it to long or to string.
Get an exception during parsing of correct JSON answer of a REST server is very, because I can't get the REST answer as string (using `vibe.web.rest`) and parse it manually.
Thanks.

### CODE REVIEW

**Comment by Geod24** [comment]
Using ulong just pushes the limit, it doesn't really fix the problem. Me think supporting `BigInt` could definitely fix this. @s-ludwig ?

File: source/vibe/data/json.d [hunk 1]
```diff
@@ -108,6 +110,7 @@ struct Json {
 		Null = null_,           /// Compatibility alias - will be deprecated soon
 		Bool = bool_,           /// Compatibility alias - will be deprecated soon
 		Int = int_,             /// Compatibility alias - will be deprecated soon
+		Uint = uint_,           /// Compatibility alias - will be deprecated soon
```
Comment (Geod24) [inline source/vibe/data/json.d]: I doubt we need to add a compatibility alias for a new feature.

**Comment by etcimon** [comment]
I also vote for a BigInt here

**Comment by s-ludwig** [comment]
`BigInt` would also have the advantage of better forward compatibility with [std_data_json](https://github.com/s-ludwig/std_data_json/). So I'd  also be in favor of that solution.

**Comment by IgorStepanov** [comment]
So, you are all prefer `BigInt`? Ok. Should it replace the long value, or there should be both types?
In the second case, how should work parsing of int values?

**Comment by s-ludwig** [comment]
It should work similar to the current `ulong` solution. So all integral numbers fitting in `long` should still be stored as a `long` and anything bigger as a `BigInt`. The only issue is that the naive implementation of this (convert `string` to `BigInt` and the `BigInt` to long if possible) will probably have a non-trivial performance impact. But for cases where that matters, we could introduce a `version (VibeJsonDisableBigInt)` that uses the old `string` -> `long` parsing mode.

**Comment by IgorStepanov** [comment]
Ok, I'll do it.

**Comment by IgorStepanov** [comment]
@s-ludwig done

File: source/vibe/data/json.d [hunk 2]
```diff
@@ -92,28 +117,59 @@ struct Json {
 		}
 	}
 
+	version (VibeUseJsonBigInt) {
+		long bigIntToLong() inout
```
Comment (s-ludwig) [inline source/vibe/data/json.d]: Missing `private`

File: source/vibe/data/json.d [hunk 3]
```diff
@@ -69,20 +68,46 @@ import std.traits;
...
+		{
+			version (VibeUseJsonBigInt)
+				BigInt bigint;
+			long long_;
+			double double_;
+			bool bool_;
+			string string_;
+			Json[string] object;
+			Json[] array;
+		}
```
Comment (s-ludwig) [inline source/vibe/data/json.d]: We need to keep the workaround unless the issues mentioned in the comment don't apply anymore.

File: source/vibe/data/json.d [hunk 4]
```diff
@@ -181,6 +253,20 @@ struct Json {
 	/// ditto
 	long opAssign(long v) { m_type = Type.int_; m_int = v; return v; }
 	/// ditto
+	version (VibeUseJsonBigInt) {
+		BigInt opAssign(BigInt v)
+		{
+			if (v >= long.min && v <= long.max) {
```
Comment (s-ludwig) [inline source/vibe/data/json.d]: I think making this choice only in the parser is a better idea. It could be surprising to have code like this trigger an error: `void test(BigInt num) { Json j = num; assert(j.type == Json.Type.bigInt); }`

**Comment by s-ludwig** [comment]
Thanks @IgorStepanov! I'm leaning towards making the `version` only apply to the parsing stage, but to keep general support for `BigInt` anyway. The current approach just seems to make it too likely to break foreign code (such as a `final switch (json.type)`). Thoughts?

Apart from my comments above, seems to look fine. I'll do a more thorough review later.

File: source/vibe/data/json.d (hunk 4 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: fixed

File: source/vibe/data/json.d (hunk 2 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: fixed

**Comment by IgorStepanov** [comment]
>I'm leaning towards making the version only apply to the parsing stage, but to keep general support for BigInt anyway. The current approach just seems to make it too likely to break foreign code (such as a final switch (json.type)). Thoughts?

Ok, I've changed the code. See the second commit.
However, If we talk only about parsing, we may simply check the number length in `skipNumber`, and construct BigInt only if it needed.
Note, JSON doesn't support non-trivial integer literals like octal, or hex thus parsing of JSON ints is so easy.
We may simply compute result while parsing a number (`num *= 10; num += (s[idx]) = '0');`) and check the overflow at each step. And no special version.

**Comment by marcioapm** [comment]
Sorry for being late to the party, but what about supporting cent/ucent, instead? :)
At first sight it seems like a more natural extension than BigInt.

I'm not sure what the state of this is in DMD but I remember it was worked on recently for LDC.

Seems like a bad spot to be in for performance sensitive scenarios - deciding between not supporting values >= 2^63 and < 2^64 or using BigInts which require extra heap storage currently, whereas the cent would fit in the Json value's buffer - converting cent back to ulong is also pretty fast.

**Comment by etcimon** [comment]
https://github.com/D-Programming-Language/dmd/blob/895eb9caf2e76208b52a48231410ce27bf37f996/src/mtype.c#L318

**Comment by IgorStepanov** [comment]
@marcioapm `cent` isn't supported in dmd and it doesn't solve the general problem: json standard allows an unlimited numbers. 
For the floating point numbers a trouble is less essential: all float types allow to represent any real number (-inf, +inf) with some error.
For an integer number json, current parser may reject a correct value which greater then long.max. I is not good.

**Comment by IgorStepanov** [comment]
@s-ludwig I've add an overflow check to `skipNumber`, and removed `version (VibeJsonDisableBigInt)`:
now I build a `BigInt` value only if number is greater then `long.max`.

File: source/vibe/data/json.d (hunk 3 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: I missed this comment, sorry. And I don't understand it. Do you tell about replacing old void*[2] payload with this union?
We don't know the BigInt size (and we can't be sure, that it will not be changed in future), thus a union, I think, is the safest choise.

File: source/vibe/data/json.d [hunk 5]
```diff
@@ -108,12 +128,12 @@ struct Json {
 		Null = null_,           /// Compatibility alias - will be deprecated soon
 		Bool = bool_,           /// Compatibility alias - will be deprecated soon
 		Int = int_,             /// Compatibility alias - will be deprecated soon
+		BigInt = bigint,       /// Compatibility alias - will be deprecated soon
```
Comment (schuetzm) [inline source/vibe/data/json.d]: This is unnecessary.

Comment (IgorStepanov) [inline source/vibe/data/json.d]: Ok, removed.

File: source/vibe/data/json.d (hunk 3 above)
Comment (s-ludwig) [inline source/vibe/data/json.d]: Yes exactly. I did the inverse some time ago to fix those issues.

You can do `void*[max((BigInt.sizeof+(void*).sizeof-1)/(void*).sizeof, 2)] m_data;` to guarantee correctness. We could try switching back to a `union`, but I know that I had severe issues with it in the past, so that we should be really sure that it doesn't reintroduce those.

Comment (IgorStepanov) [inline source/vibe/data/json.d]: Why `void*`?
Why not `ubyte[max(BigInt.sizeof, (void*).sizeof * 2)] m_data;`?

Comment (IgorStepanov) [inline source/vibe/data/json.d]: Ok, I've changed it as you say. Is it ok now?
BTW, what troubles with union do you have met?

Comment (IgorStepanov) [inline source/vibe/data/json.d]: ldc raises a strange error with `void*[]` inside biguintcore. See test results.
I'm trying `ubyte[]`, but I think, it will fail too.

Comment (IgorStepanov) [inline source/vibe/data/json.d]: Ok, I've added explicit initialization of m_bigint before assigning. Now all works.
Rebased.
Is it ok now?

**Comment by Geod24** [comment]
Could you squash your commits into a single one ?

File: examples/json/source/app.d [hunk 6]
```diff
@@ -17,4 +17,12 @@ void main()
 	Json parent = obj;
 	parent.remove("item1");
 	foreach (i; obj) writeln(i);
+
+	version (VibeJsonDisableBigInt) {
+	} else {
```
Comment (Geod24) [inline examples/json/source/app.d]: This seem to be the only occurence of the version left.

File: examples/json/source/app.d [hunk 7]
```diff
@@ -17,4 +17,12 @@ void main()
 	Json parent = obj;
 	parent.remove("item1");
 	foreach (i; obj) writeln(i);
+
+	version (VibeJsonDisableBigInt) {
+	} else {
+		import std.bigint;
+		auto obj2 = parseJsonString(`{"serial":17559991181826658461}`);
+		writeln("serial: ", obj2["serial"]);
+		assert(obj2["serial"] == BigInt(17559991181826658461UL));
```
Comment (Geod24) [inline examples/json/source/app.d]: Also, it'd be good to have a test that test number outside of a size_t boundary, to ensure that this part is also working.

File: source/vibe/data/json.d [hunk 8]
```diff
@@ -44,7 +44,7 @@ unittest {
 public import vibe.data.serialization;
 
 public import std.json : JSONException;
-import std.algorithm : equal, min;
+import std.algorithm : equal, min, max;
```
Comment (Geod24) [inline source/vibe/data/json.d]: Minor: I would suggest that we reduce, and even remove, any selective import at the top. Until 313 and 314 are fixed, those cause symbols leakage, and upgrading a version of Vibe.d which removes one of the import might break code. And when 313 / 314 are fixed, it'll definitely breaks a lot of code...

File: source/vibe/data/json.d [hunk 9]
```diff
@@ -69,14 +69,23 @@ import std.traits;
...
 		// because of internal pointers. This crude workaround seems to fix
 		// the issues.
-		void*[2] m_data;
+		void*[max((BigInt.sizeof+(void*).sizeof-1)/(void*).sizeof, 2)] m_data;
 		ref inout(T) getDataAs(T)() inout { static assert(T.sizeof <= m_data.sizeof); return *cast(inout(T)*)m_data.ptr; }
+
+		void initBigInt()
+		{
+			BigInt init_;
+			(cast(ubyte*)m_data.ptr)[0 .. BigInt.sizeof] = (cast(ubyte*)&init_)[0 .. BigInt.sizeof];
```
Comment (Geod24) [inline source/vibe/data/json.d]: So that's the LDC bug I guess. Do you have an issue number, or something ? I think anyone reading this in a 6+ months time would be like "WHAT? WHY?".

**Comment by IgorStepanov** [comment]
>Could you squash your commits into a single one ?

All commits represents different approaches. If the my last approach will be considered best, I squash commits. Otherwice, it will be much easy to rollback to previous version.

File: source/vibe/data/json.d [hunk 10]
```diff
@@ -181,6 +199,8 @@ struct Json {
 	/// ditto
 	long opAssign(long v) { m_type = Type.int_; m_int = v; return v; }
 	/// ditto
+	BigInt opAssign(BigInt v) { m_type = Type.bigint; initBigInt(); m_bigint = v; return v; }
```
Comment (Geod24) [inline source/vibe/data/json.d]: IDK how `BigInt` are implemented, but I can see the cycle: 'blitting the default value' -> 'assigning data' -> 'Do it again' as potentially problematic: if `BigInt` use `malloc`ed memory, it's more likely that on assignement, either `malloc` or `free` (or both) happen.

File: source/vibe/data/json.d (hunk 8 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: Do you suggest to move `std.algorithm : max` inside struct Json?
I can't move/remove `import std.algorithm : equal, min` because it irrelevant with this PR.

File: examples/json/source/app.d (hunk 7 above)
Comment (IgorStepanov) [inline examples/json/source/app.d]: Ok. Please wait few minutes.

File: examples/json/source/app.d (hunk 6 above)
Comment (IgorStepanov) [inline examples/json/source/app.d]: My bad. I'll remove it.

**Comment by Geod24** [comment]
> All commits represents different approaches. If the my last approach will be considered best, I squash commits. Otherwice, it will be much easy to rollback to previous version.

That request is just to make the history easier to understand. I often find myself `blame`ing a file, to see when the code was introduced, and why. Having to go through various modifications for the same P.R. just eats a lot of time (same goes for history).

As per the rollback, with git, you can easily checkout a branch that points to your current commit, and squash them together. `git` never destroys old objects - unless nothing points to it, and they are gc'ed, which is ~ 2 weeks later, but won't happen with another branch -.

File: source/vibe/data/json.d (hunk 8 above)
Comment (Geod24) [inline source/vibe/data/json.d]: I'll just make it `import std.algorithm` and call it a day.

File: source/vibe/data/json.d (hunk 9 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: I think it is not a bug. Simply dmd and gdc uses another BigInt implementation: with inline asm.
The general idea: BigInt.init contains a BigUInt instance, which initially should contatin a [0] array.
When we get BigInt via `getDataAs`, it may contain the garbarge (null or something worse) (and BigUInt invariant is fault). Thus, we should explicitly write BigInt.init to storage before the first using.

File: source/vibe/data/json.d (hunk 10 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: >IDK how BigInt are implemented, but I can see the cycle

Default value may be only static value, because BigInt is a struct, and BigInt.init should be known at compile time.
Thus everytime, `a`, `b` and `c` in the following example has a identical (bitwise) value. Thus I simply write BigInt.init to the m_data.
```
    BigInt a;
    BigInt b;
    BigInt c;
```

File: source/vibe/data/json.d (hunk 9 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: >I think anyone reading this in a 6+ months time would be like "WHAT? WHY?".

I should add the comment, yes?

File: source/vibe/data/json.d (hunk 8 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: done

File: examples/json/source/app.d (hunk 7 above)
Comment (IgorStepanov) [inline examples/json/source/app.d]: done

File: examples/json/source/app.d (hunk 6 above)
Comment (IgorStepanov) [inline examples/json/source/app.d]: done

File: source/vibe/data/json.d (hunk 10 above)
Comment (Geod24) [inline source/vibe/data/json.d]: What I was refering to was, if someone calls `opAssign` with a `BigInt` value multiple time, the finalizers of all but the last 'instance' will not be run.

Comment (IgorStepanov) [inline source/vibe/data/json.d]: This case is fixed.
`initBigInt()` is called only if old `m_type` in assign is not `Type.bigint`

File: source/vibe/data/json.d (hunk 9 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: Comments have been added.

File: source/vibe/data/json.d (hunk 8 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: `/tests/redis/source/app.d` used equal from `json`. I've added an import to app.d

File: source/vibe/data/json.d [hunk 11]
```diff
@@ -181,6 +209,15 @@ struct Json {
 	/// ditto
 	long opAssign(long v) { m_type = Type.int_; m_int = v; return v; }
 	/// ditto
+	BigInt opAssign(BigInt v)
+	{
+		if (m_type != Type.bigint)
+			initBigInt();
```
Comment (Geod24) [inline source/vibe/data/json.d]: IIUC, in the following code:
````
Json bi;
bi = BigInt("42");
bi = 42L;
bi = BigInt("42");
````

The bug is still present, isn't it ?

**Comment by s-ludwig** [comment]
@IgorStepanov WRT `union` vs `void*[...]`, the issues I had were memory leaks, but also it broke using `std.algorithm.swap` (and thus `std.sort` etc.). `void*` needs to be used instead of `ubyte` to tell the GC to look for pointers within the array. If we'd store for example a `Json[]` within an `ubyte[]` array, it may simply get collected even if it's still referenced.

**Comment by s-ludwig** [comment]
@Geod24 The destructor issue is actually a good point. In fact, the same issue is in [std_data_json/`JSONNumber`](https://github.com/s-ludwig/std_data_json/blob/066eacbbeeec13dd437466ad62ea3d92e7f3ee46/source/stdx/data/json/lexer.d#L1159) and nobody has noticed it yet (probably because it looks so natural with the `union` syntax). Maybe it would be worth to at least add a warning when a type with destructor is used within a `union` and that union is stored in a struct without destructors (same for postblit and possibly opAssign).

Also, I'll have to find the tagged union implementation for Phobos again that I had started some time ago and submit a PR. `Algebraic` is (not so) nice for many things, but often it's simply desirable to work with a type `enum` instead of `TypeInfo`s and one shouldn't have to mess with this kind of low level stuff in regular code.

File: source/vibe/data/json.d (hunk 11 above)
Comment (IgorStepanov) [inline source/vibe/data/json.d]: Ok, I've fixed it. (see the below description)

**Comment by IgorStepanov** [comment]
@s-ludwig @Geod24 I've fixed BigInt destruct issue.
Now when we try to call `opAssign` for BigInt objects, and save non-BigInt to it, `finiBigInt` is called.
`finiBigInt` simply creates an automatic `BigInt init_` variable, and `swap`s it with the stored BigInt value.
When `finiBigInt` is finished, old BigInt value will be destroyed with `init_`, and `m_bigint` now contains a `BigInt.init` value, which can be rewritten with any non-BigInt value without any possible leaks.

**Comment by s-ludwig** [comment]
Looks good to me now. I'll just make some minor, mostly cosmetic, adjustments after merge. Does anyone else still see any issue?

**Comment by Geod24** [comment]
Well... I look forward to `std.data.json` :D

**Comment by s-ludwig** [comment]
Good thing is that this change improves compatibility with `std.data.json`.

Merging now. Thanks Igor for your work!

**Comment by s-ludwig** [comment]
Okay, got the changes in. The only interface change has been to rename `Type.bigint` to `Type.bigInt`.

**Comment by IgorStepanov** [comment]
>Okay, got the changes in. The only interface change has been to rename Type.bigint to Type.bigInt.

Thanks! 
BTW, when do you plan to release the next vibe.d version (with this patch)? I mean, that when I'll able to fetch it with `dub`.
I use vibe.d in a commercial product, and if the next release will not be soon, I'll write temporary code to handle this issue until the next version will be released.

**Comment by s-ludwig** [comment]
I could tag an alpha/beta release right away.

**Comment by s-ludwig** [comment]
Okay 0.7.24-beta.1 is tagged now.

**Comment by IgorStepanov** [comment]
>Okay 0.7.24-beta.1 is tagged now.

Thanks a lot! :)

**Comment by MartinNowak** [comment]
???
Javascript doesn't have integer types at all, it uses double-precision (64-bit) float for any Number.
[math - What is JavaScript's highest integer value that a Number can go to without losing precision? - Stack Overflow](http://stackoverflow.com/questions/307179/what-is-javascripts-highest-integer-value-that-a-number-can-go-to-without-losin)
Try `alert(01234567890123456789);`.

**Comment by s-ludwig** [comment]
I don't think that this really matters, since this is JSON and not JavaScript. JSON doesn't specify anything about the numerical representation, so the implementation is basically free to choose whatever fits. Since large integers are used in practice, it would be quite arbitrary to not support them.
